### ISSUE TITLE
FTP vs WFS

### ISSUE THREAD

**Issue by smnorris** [smnorris/fwapg#127]
It looks like FWA FTP data is being updated monthly.
ftp://ftp.geobc.gov.bc.ca/sections/outgoing/bmgs/FWA_Public

If this is the case, it will be worth migrating back to using this as the data source, downloading the archive should be more efficient and reliable than via `bcdata`/ WFS

**Comment by NewGraphEnvironment** [comment]
Keep getting kicked out of the `bcdata bc2pg` calls with what I think are socket.timeout errors. Running on `s-4vcpu-8gb-intel` droplet.  Do you think it is a WFS thing, have any advise or do you know of a setting that I might be able to change to avoid it?

Since you have the handy `make` thing happening I am looping over the `make` command with 

`for x in {1..20}; do make; done` ha

Below is the error report


```

psql postgresql://airvine@localhost:5432/newgraph -v ON_ERROR_STOP=1 -c "delete from fwapg.fwa_linear_boundaries_sp where watershed_group_code = 'LCHL'"
DELETE 0
# load to staging schema
bcdata bc2pg --db_url postgresql://airvine@localhost:5432/newgraph -t \
	--schema fwapg \
	--query "WATERSHED_GROUP_CODE = 'LCHL'" \
	--append \
	whse_basemapping.fwa_linear_boundaries_sp
INFO:bcdata.wfs:Total features requested: 18644
Traceback (most recent call last):
  File "/home/airvine/miniconda3/envs/bcfishpass/lib/python3.9/site-packages/urllib3/connectionpool.py", line 449, in _make_request
    six.raise_from(e, None)
  File "<string>", line 3, in raise_from
  File "/home/airvine/miniconda3/envs/bcfishpass/lib/python3.9/site-packages/urllib3/connectionpool.py", line 444, in _make_request
    httplib_response = conn.getresponse()
  File "/home/airvine/miniconda3/envs/bcfishpass/lib/python3.9/http/client.py", line 1377, in getresponse
    response.begin()
  File "/home/airvine/miniconda3/envs/bcfishpass/lib/python3.9/http/client.py", line 320, in begin
    version, status, reason = self._read_status()
  File "/home/airvine/miniconda3/envs/bcfishpass/lib/python3.9/http/client.py", line 281, in _read_status
    line = str(self.fp.readline(_MAXLINE + 1), "iso-8859-1")
  File "/home/airvine/miniconda3/envs/bcfishpass/lib/python3.9/socket.py", line 704, in readinto
    return self._sock.recv_into(b)
socket.timeout: timed out

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/airvine/miniconda3/envs/bcfishpass/lib/python3.9/site-packages/requests/adapters.py", line 489, in send
    resp = conn.urlopen(
  File "/home/airvine/miniconda3/envs/bcfishpass/lib/python3.9/site-packages/urllib3/connectionpool.py", line 787, in urlopen
    retries = retries.increment(
  File "/home/airvine/miniconda3/envs/bcfishpass/lib/python3.9/site-packages/urllib3/util/retry.py", line 550, in increment
    raise six.reraise(type(error), error, _stacktrace)
  File "/home/airvine/miniconda3/envs/bcfishpass/lib/python3.9/site-packages/urllib3/packages/six.py", line 770, in reraise
    raise value
  File "/home/airvine/miniconda3/envs/bcfishpass/lib/python3.9/site-packages/urllib3/connectionpool.py", line 703, in urlopen
    httplib_response = self._make_request(
  File "/home/airvine/miniconda3/envs/bcfishpass/lib/python3.9/site-packages/urllib3/connectionpool.py", line 451, in _make_request
    self._raise_timeout(err=e, url=url, timeout_value=read_timeout)
  File "/home/airvine/miniconda3/envs/bcfishpass/lib/python3.9/site-packages/urllib3/connectionpool.py", line 340, in _raise_timeout
    raise ReadTimeoutError(
urllib3.exceptions.ReadTimeoutError: HTTPConnectionPool(host='openmaps.gov.bc.ca', port=80): Read timed out. (read timeout=30)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/airvine/miniconda3/envs/bcfishpass/bin/bcdata", line 8, in <module>
    sys.exit(cli())
  File "/home/airvine/miniconda3/envs/bcfishpass/lib/python3.9/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "/home/airvine/miniconda3/envs/bcfishpass/lib/python3.9/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "/home/airvine/miniconda3/envs/bcfishpass/lib/python3.9/site-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/airvine/miniconda3/envs/bcfishpass/lib/python3.9/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/airvine/miniconda3/envs/bcfishpass/lib/python3.9/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "/home/airvine/miniconda3/envs/bcfishpass/lib/python3.9/site-packages/bcdata/cli.py", line 357, in bc2pg
    out_table = bcdata.bc2pg(
  File "/home/airvine/miniconda3/envs/bcfishpass/lib/python3.9/site-packages/bcdata/bc2pg.py", line 79, in bc2pg
    urls = bcdata.define_requests(
  File "/home/airvine/miniconda3/envs/bcfishpass/lib/python3.9/site-packages/bcdata/wfs.py", line 151, in define_requests
    wfs = WebFeatureService(url=bcdata.OWS_URL, version="2.0.0")
  File "/home/airvine/miniconda3/envs/bcfishpass/lib/python3.9/site-packages/owslib/wfs.py", line 56, in WebFeatureService
    return wfs200.WebFeatureService_2_0_0(
  File "/home/airvine/miniconda3/envs/bcfishpass/lib/python3.9/site-packages/owslib/feature/wfs200.py", line 128, in __init__
    self._capabilities = reader.read(self.url, self.timeout)
  File "/home/airvine/miniconda3/envs/bcfishpass/lib/python3.9/site-packages/owslib/feature/common.py", line 54, in read
    u = openURL(request, timeout=timeout, headers=self.headers, auth=self.auth)
  File "/home/airvine/miniconda3/envs/bcfishpass/lib/python3.9/site-packages/owslib/util.py", line 208, in openURL
    req = requests.request(method.upper(), url_base, headers=headers, **rkwargs)
  File "/home/airvine/miniconda3/envs/bcfishpass/lib/python3.9/site-packages/requests/api.py", line 59, in request
    return session.request(method=method, url=url, **kwargs)
  File "/home/airvine/miniconda3/envs/bcfishpass/lib/python3.9/site-packages/requests/sessions.py", line 587, in request
    resp = self.send(prep, **send_kwargs)
  File "/home/airvine/miniconda3/envs/bcfishpass/lib/python3.9/site-packages/requests/sessions.py", line 701, in send
    r = adapter.send(request, **kwargs)
  File "/home/airvine/miniconda3/envs/bcfishpass/lib/python3.9/site-packages/requests/adapters.py", line 578, in send
    raise ReadTimeout(e, request=request)
requests.exceptions.ReadTimeout: HTTPConnectionPool(host='openmaps.gov.bc.ca', port=80): Read timed out. (read timeout=30)
make: *** [Makefile:104: .make/fwa_linear_boundaries_sp_LCHL] Error 1
```

**Comment by smnorris** [comment]
I have been seeing that error a lot too. Perhaps there is more wfs usage these days or a setting has changed. Unfortunately there is no bcdata setting to compensate. A couple of options:- add some kind of check to bcdata to avoid the timeouts- download FWA data from FTP rather than WFS (this is preferred and far more efficient- but I have avoided because updates to FTP were inconsistent and on one instance it was broken. GeoBC is now manually keeping it current so this is worth revisiting)On Jan 26, 2023, at 8:10 AM, Allan Irvine ***@***.***> wrote:﻿
Keep getting kicked out of the bcdata bc2pg calls with what I think are socket.timeout errors. Running on s-4vcpu-8gb-intel droplet.  Do you think it is a WFS thing, have any advise or do you know of a setting that I might be able to change to avoid it?
Since you have the handy make thing happening I am looping over the make command with
for x in {1..20}; do make; done ha
Below is the error report

psql ***@***.***:5432/newgraph -v ON_ERROR_STOP=1 -c "delete from fwapg.fwa_linear_boundaries_sp where watershed_group_code = 'LCHL'"
DELETE 0
# load to staging schema
bcdata bc2pg --db_url ***@***.***:5432/newgraph -t \
	--schema fwapg \
	--query "WATERSHED_GROUP_CODE = 'LCHL'" \
	--append \
	whse_basemapping.fwa_linear_boundaries_sp
INFO:bcdata.wfs:Total features requested: 18644
Traceback (most recent call last):
  File "/home/airvine/miniconda3/envs/bcfishpass/lib/python3.9/site-packages/urllib3/connectionpool.py", line 449, in _make_request
    six.raise_from(e, None)
  File "<string>", line 3, in raise_from
  File "/home/airvine/miniconda3/envs/bcfishpass/lib/python3.9/site-packages/urllib3/connectionpool.py", line 444, in _make_request
    httplib_response = conn.getresponse()
  File "/home/airvine/miniconda3/envs/bcfishpass/lib/python3.9/http/client.py", line 1377, in getresponse
    response.begin()
  File "/home/airvine/miniconda3/envs/bcfishpass/lib/python3.9/http/client.py", line 320, in begin
    version, status, reason = self._read_status()
  File "/home/airvine/miniconda3/envs/bcfishpass/lib/python3.9/http/client.py", line 281, in _read_status
    line = str(self.fp.readline(_MAXLINE + 1), "iso-8859-1")
  File "/home/airvine/miniconda3/envs/bcfishpass/lib/python3.9/socket.py", line 704, in readinto
    return self._sock.recv_into(b)
socket.timeout: timed out

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/airvine/miniconda3/envs/bcfishpass/lib/python3.9/site-packages/requests/adapters.py", line 489, in send
    resp = conn.urlopen(
  File "/home/airvine/miniconda3/envs/bcfishpass/lib/python3.9/site-packages/urllib3/connectionpool.py", line 787, in urlopen
    retries = retries.increment(
  File "/home/airvine/miniconda3/envs/bcfishpass/lib/python3.9/site-packages/urllib3/util/retry.py", line 550, in increment
    raise six.reraise(type(error), error, _stacktrace)
  File "/home/airvine/miniconda3/envs/bcfishpass/lib/python3.9/site-packages/urllib3/packages/six.py", line 770, in reraise
    raise value
  File "/home/airvine/miniconda3/envs/bcfishpass/lib/python3.9/site-packages/urllib3/connectionpool.py", line 703, in urlopen
    httplib_response = self._make_request(
  File "/home/airvine/miniconda3/envs/bcfishpass/lib/python3.9/site-packages/urllib3/connectionpool.py", line 451, in _make_request
    self._raise_timeout(err=e, url=url, timeout_value=read_timeout)
  File "/home/airvine/miniconda3/envs/bcfishpass/lib/python3.9/site-packages/urllib3/connectionpool.py", line 340, in _raise_timeout
    raise ReadTimeoutError(
urllib3.exceptions.ReadTimeoutError: HTTPConnectionPool(host='openmaps.gov.bc.ca', port=80): Read timed out. (read timeout=30)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/airvine/miniconda3/envs/bcfishpass/bin/bcdata", line 8, in <module>
    sys.exit(cli())
  File "/home/airvine/miniconda3/envs/bcfishpass/lib/python3.9/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "/home/airvine/miniconda3/envs/bcfishpass/lib/python3.9/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "/home/airvine/miniconda3/envs/bcfishpass/lib/python3.9/site-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/airvine/miniconda3/envs/bcfishpass/lib/python3.9/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/airvine/miniconda3/envs/bcfishpass/lib/python3.9/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "/home/airvine/miniconda3/envs/bcfishpass/lib/python3.9/site-packages/bcdata/cli.py", line 357, in bc2pg
    out_table = bcdata.bc2pg(
  File "/home/airvine/miniconda3/envs/bcfishpass/lib/python3.9/site-packages/bcdata/bc2pg.py", line 79, in bc2pg
    urls = bcdata.define_requests(
  File "/home/airvine/miniconda3/envs/bcfishpass/lib/python3.9/site-packages/bcdata/wfs.py", line 151, in define_requests
    wfs = WebFeatureService(url=bcdata.OWS_URL, version="2.0.0")
  File "/home/airvine/miniconda3/envs/bcfishpass/lib/python3.9/site-packages/owslib/wfs.py", line 56, in WebFeatureService
    return wfs200.WebFeatureService_2_0_0(
  File "/home/airvine/miniconda3/envs/bcfishpass/lib/python3.9/site-packages/owslib/feature/wfs200.py", line 128, in __init__
    self._capabilities = reader.read(self.url, self.timeout)
  File "/home/airvine/miniconda3/envs/bcfishpass/lib/python3.9/site-packages/owslib/feature/common.py", line 54, in read
    u = openURL(request, timeout=timeout, headers=self.headers, auth=self.auth)
  File "/home/airvine/miniconda3/envs/bcfishpass/lib/python3.9/site-packages/owslib/util.py", line 208, in openURL
    req = requests.request(method.upper(), url_base, headers=headers, **rkwargs)
  File "/home/airvine/miniconda3/envs/bcfishpass/lib/python3.9/site-packages/requests/api.py", line 59, in request
    return session.request(method=method, url=url, **kwargs)
  File "/home/airvine/miniconda3/envs/bcfishpass/lib/python3.9/site-packages/requests/sessions.py", line 587, in request
    resp = self.send(prep, **send_kwargs)
  File "/home/airvine/miniconda3/envs/bcfishpass/lib/python3.9/site-packages/requests/sessions.py", line 701, in send
    r = adapter.send(request, **kwargs)
  File "/home/airvine/miniconda3/envs/bcfishpass/lib/python3.9/site-packages/requests/adapters.py", line 578, in send
    raise ReadTimeout(e, request=request)
requests.exceptions.ReadTimeout: HTTPConnectionPool(host='openmaps.gov.bc.ca', port=80): Read timed out. (read timeout=30)
make: *** [Makefile:104: .make/fwa_linear_boundaries_sp_LCHL] Error 1

**Comment by smnorris** [comment]
fixed with v0.4.0
